File: client/scratch/smooth_follow/springs.py
from dataclasses import InitVar, dataclass
from math import ceil
from typing import ClassVar, Union


@dataclass
class SpringConfig:
    tension: float = 170
    friction: float = 26
    mass: float = 1
    precision: float = None
    velocity: float = None
    clamp: bool = False
    bounce: float = None

    @classmethod
    def gentle(cls):
        return SpringConfig(tension=120, friction=14)

    @classmethod
    def wobbly(cls):
        return SpringConfig(tension=180, friction=12)

    @classmethod
    def stiff(cls):
        return SpringConfig(tension=210, friction=20)

    @classmethod
    def slow(cls):
        return SpringConfig(tension=280, friction=60)

    @classmethod
    def molasses(cls):
        return SpringConfig(tension=280, friction=120)


@dataclass
class AnimatedValue:
    cur_pos: float = None
    from_pos: float = 0
    to_pos: float = 0
    last_vel: float = None

    def __post_init__(self):
        self.cur_pos = self.cur_pos or self.from_pos


def step_spring(
    dt_secs: float, val: AnimatedValue, config: SpringConfig = SpringConfig()
):
    finished = False
    v0 = config.velocity or 0

    precision = config.precision or (
        0.005
        if val.from_pos == val.to_pos
        else min(1, abs(val.to_pos - val.from_pos) * 0.001)
    )
    velocity = val.last_vel or v0

    # The velocity at which movement is essentially none
    rest_velocity = precision / 10

    # Bouncing is opt-in (not to be confused with overshooting)
    bounce_factor = 0 if config.clamp else config.bounce
    can_bounce = bounce_factor != 0

    # When `True`, the value is increasing over time
    is_growing = v0 > 0 if val.from_pos == val.to_pos else val.from_pos < val.to_pos

    # When `True`, the velocity is considered moving
    is_moving: bool

    # When `True`, the velocity is being deflected or clamped
    is_bouncing = False

    step = 1  # 1ms
    num_steps = ceil(dt_secs / (step / 1000))
    for _ in range(num_steps):
        is_moving = abs(velocity) > rest_velocity

        if not is_moving:
            finished = abs(val.to_pos - val.cur_pos) <= precision
            if finished:
                break

        if can_bounce:
            is_bouncing = (
                val.cur_pos == val.to_pos or val.cur_pos > val.to_pos == is_growing
            )

            # Invert the velocity with a magnitude, or clamp it.
            if is_bouncing:
                velocity = -velocity * bounce_factor
                val.cur_pos = val.to_pos

        spring_force = -config.tension * 0.000001 * (val.cur_pos - val.to_pos)
        damping_force = -config.friction * 0.001 * velocity
        acceleration = (spring_force + damping_force) / config.mass  # pt/ms^2

        velocity = velocity + acceleration * step  # pt/ms
        val.cur_pos = val.cur_pos + velocity * step

    val.last_vel = velocity
    return finished


if __name__ == "__main__":
    val = AnimatedValue(from_pos=200, to_pos=640)
    config = SpringConfig.gentle()

    for frame in range(500):
        if step_spring(dt_secs=16 / 1000, val=val, config=config):
            print(f"finished on frame {frame}")
            break

        print(val.cur_pos)

File: hal/src/motion/springs.rs
pub struct SpringConfig {
    pub tension: f32,
    pub friction: f32,
    pub mass: f32,
    pub precision: Option<f32>,
    pub velocity: Option<f32>,
    pub clamp: bool,
    pub bounce: Option<f32>,
}

impl Default for SpringConfig {
    fn default() -> Self {
        Self {
            tension: 170.0,
            friction: 26.0,
            mass: 1.0,
            precision: None,
            velocity: None,
            clamp: false,
            bounce: None,
        }
    }
}

impl SpringConfig {
    pub fn gentle() -> Self {
        Self {
            tension: 120.0,
            friction: 14.0,
            ..Default::default()
        }
    }

    pub fn wobbly() -> Self {
        Self {
            tension: 180.0,
            friction: 12.0,
            ..Default::default()
        }
    }

    pub fn stiff() -> Self {
        Self {
            tension: 210.0,
            friction: 20.0,
            ..Default::default()
        }
    }

    pub fn slow() -> Self {
        Self {
            tension: 280.0,
            friction: 60.0,
            ..Default::default()
        }
    }

    pub fn molasses() -> Self {
        Self {
            tension: 280.0,
            friction: 120.0,
            ..Default::default()
        }
    }
}

struct Spring {
    current_value: f32,
    from_value: f32,
    to_value: f32,
    last_velocity: Option<f32>,
}

// def step_spring(
//     dt_secs: float, val: AnimatedValue, config: SpringConfig = SpringConfig()
// ):
fn update_spring(s: &mut Spring, config: SpringConfig) {
    let mut finished = false;
    let v0 = config.velocity.unwrap_or(0.0);
    let precision = config.precision.unwrap_or_else(|| {
        if s.from_value == s.to_value {
            0.005
        } else {
            1.0_f32.min((s.to_value - s.from_value).abs() * 0.001)
        }
    });
    let mut velocity = s.last_velocity.unwrap_or(v0);

    // The velocity at which movement is essentially none
    let rest_velocity = precision / 10.0;

    // Bouncing is opt-in (not to be confused with overshooting)
    let bounce_factor = if config.clamp {
        0.0
    } else {
        config.bounce.unwrap_or(0.0)
    };
    let can_bounce = bounce_factor != 0.0;

    // When `true`, the value is increasing over time
    let is_growing = if s.from_value == s.to_value {
        v0 > 0.0
    } else {
        s.from_value < s.to_value
    };

    // When `true`, the velocity is considered moving
    let mut is_moving: bool;

    // When `true`, the velocity is being deflected or clamped
    let mut is_bouncing;

    let step = 1; // 1ms
    let dt_secs = 1.4214f32;
    let num_steps = (dt_secs / (step as f32 / 1000.0)).ceil() as u32;
    for _ in 0..num_steps {
        is_moving = velocity.abs() > rest_velocity;

        if !is_moving {
            finished = (s.to_value - s.current_value).abs() <= precision;
            if finished {
                break;
            }
        }

        if can_bounce {
            is_bouncing =
                s.current_value == s.to_value || (s.current_value > s.to_value) == is_growing;

            // Invert the velocity with a magnitude, or clamp it.
            if is_bouncing {
                velocity = -velocity * bounce_factor;
                s.current_value = s.to_value;
            }
        }
        let spring_force = -config.tension * 0.000001 * (s.current_value - s.to_value);
        let damping_force = -config.friction * 0.001 * velocity;
        let acceleration = (spring_force + damping_force) / config.mass; // pt/ms^2

        velocity = velocity + acceleration * step as f32; // pt/ms
        s.current_value = s.current_value + velocity * step as f32;
    }
    s.last_velocity = Some(velocity);
    // return finished
}
